### DISCUSSION TITLE
Allow configuring `pgaudit.log_parameter`

### DISCUSSION THREAD

**Discussion by xEverth** [supabase/supabase#20183] [Feature Requests]
pgaudit is the recommended method to log according to the docs: https://supabase.com/docs/guides/database/extensions/pgaudit
It's possible to enable logging (for a session, or for specific roles) and to choose the log level: https://supabase.com/docs/guides/platform/logs#configuring-pgauditlog

I'd like to verify the call parameters of the RPCs called via postgREST, but I am not able to alter pgaudit.log_parameter due to a permission error: 

> ERROR: permission denied to set parameter "pgaudit.log_parameter"

## To Reproduce

Connect to the database as postgres or use the SQL Editor from the dashboard and run either of the following:
set pgaudit.log_parameter to 'on';
alter role authenticator set pgaudit.log_parameter to 'on';

## Expected behavior

I expect to be able to set log_parameter and audit the parameters sent to the RPC.
https://github.com/pgaudit/pgaudit?tab=readme-ov-file#pgauditlog_parameter

## Screenshots
![image](https://github.com/supabase/supabase/assets/8996551/64faab25-37a0-4793-9e3f-054af29c6744)
![image](https://github.com/supabase/supabase/assets/8996551/1b89edc9-5751-44af-aaaa-366da8ba495c)
![image](https://github.com/supabase/supabase/assets/8996551/f07d2a80-55c2-4687-820f-24ed9ddb3262)

## Additional context
Postgres version: 15.1.0.113
PostgREST version: 11.2.0

**Comment by xEverth** [comment]
I'll note that I found a way to access the (URL Encoded) parameters for the RPCs from the API Edge Network Logs > Request Metadata > Search. It's not exactly what I was looking for, but it's something!

**Comment by soedirgo** [comment]
Hi @xEverth! Currently configuring `pgaudit.log_parameter` is not supported, since it renders [pgsodium](https://supabase.com/docs/guides/database/extensions/pgsodium) useless (secrets in encrypted columns will be logged unencrypted). This may not be relevant if you're not using pgsodium or [Supabase Vault](https://supabase.com/docs/guides/database/vault).

One workaround for this is using [triggers](https://github.com/orgs/supabase/discussions/17758).

I've converted this issue to a feature request to gauge interest - if enough people want this restriction lifted we may reconsider removing this restriction.

**Reply by brandonwheat** [reply]
would love this feature for temporary troubleshooting

**Reply by evelant** [reply]
Would also love to be able to use this for troubleshooting. We've got a very difficult problem taking our whole instance down at random. There's not enough information in the logs for us to diagnose. We don't use pgsodium or vault so we're not concerned about leaking any encrypted data to logs.

**Comment by ncryptedV1** [comment]
@soedirgo sadly, the trigger-based approach doesn't allow for logging reads (SELECT), and without parameter binding, it is impossible to be compliant with the strict auditing requirements over here in Europe (know which specific records/rows have been accessed) - especially in healthcare. We're self-hosting, and while `log_parameter` can be freely set there, it has absolutely no effect and is overwritten when starting the session for the affected user → I guess you guys patched pgaudit. Is there any way I can bypass this at least on my self-hosted version, other than building my own image or doing nasty patches?

**Reply by soedirgo** [reply]
If you're self-hosting, you should be able to modify it by connecting as `supabase_admin`

**Reply by ncryptedV1** [reply]
Worked, thanks! What's the technical explanation for this? Even with the postgres role, I got no permission error, and the setting was persistent across sessions and restarts globally:
```
SELECT 
    r.rolname                                           AS role_name,
    COALESCE(d.datname, 'GLOBAL')                       AS scope,
    unnest(COALESCE(s.setconfig, r.rolconfig))          AS raw_setting,
    regexp_match(unnest(COALESCE(s.setconfig, r.rolconfig)), '^pgaudit\.([^=]+)=(.*)') AS parsed
FROM pg_roles r
LEFT JOIN pg_db_role_setting s ON s.setrole = r.oid
LEFT JOIN pg_database d        ON d.oid = s.setdatabase
WHERE (r.rolconfig::text ILIKE '%pgaudit%' 
    OR s.setconfig::text   ILIKE '%pgaudit%')
ORDER BY role_name, scope;
```
→ looks exactly the same after regardless of whether I set it with postgres or supabase_admin
| role_name     | scope  | raw_setting                          | parsed                        |
| ------------- | ------ | ------------------------------------ | ----------------------------- |
...
| authenticator | GLOBAL | pgaudit.log_parameter=on             | ["log_parameter","on"]        |
...

**Reply by soedirgo** [reply]
~Ah, I assumed from your original comment that you tried running it as `postgres` but it didn't work. It actually should work that way on the hosted platform, but hadn't tested it on the self-hosted setup.~

Correction: `pgaudit.log_parameter` isn't configurable on the hosted platform either. I think it works on the self-hosted platform because `postgres` is a superuser there, so it won't hit permission errors

**Reply by ncryptedV1** [reply]
Yes, but it does show the parameters now after setting it with supabase_admin. Maybe some overnight replication magic happened, but I restarted multiple times after setting it with postgres and it continuously showed `,<not logged>` in the audit logs normal for `log_parameter=off` (which were `'on'` though) up until setting it with supabase_admin
